core/trino-main/src/main/java/io/trino/operator/aggregation/partial/PartialAggregationControlFactory.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.aggregation.partial;
+
+import io.trino.Session;
+import io.trino.operator.aggregation.partial.PartialAggregationControl.PartialAggregationTracker;
+import io.trino.sql.planner.plan.AggregationNode;
+
+import static io.trino.SystemSessionProperties.getAdaptivePartialAggregationMinRows;
+import static io.trino.SystemSessionProperties.getAdaptivePartialAggregationUniqueRowsRatio;
+import static io.trino.SystemSessionProperties.isAdaptivePartialAggregationEnabled;
+
+public class PartialAggregationControlFactory
```

Comment:
Let's remove factory and `NoOpPartialAggregationControl` and just make `Optional<PartialAggregationController> partialAggregationControler` in `HashAggregationOperatorFactory`

Reply:
> What is the benefit from Optional<PartialAggregationController>?

If you use `Optional` then you don't have to have `PartialAggregationControl` or `PartialAggregationControlFactory` as interfaces (since there is just single implementation). There isn't really going to be another implementation and it won't be pluggable really. Hence, an interface just to have noop is an overkill.